src/py2spack/package_providers.py

```diff
@@ -0,0 +1,316 @@
+"""Utils for downloading packages from PyPI and opening the pyproject.tomls."""
+
+from __future__ import annotations
+
+import abc
+import functools
+from typing import Any, Protocol
+
+import requests
+from packaging import version as vn
+
+from py2spack import utils
+
+
+KNOWN_ARCHIVE_FORMATS = [
+    ".tar",
+    ".tar.gz",
+    ".tar.bz2",
+    ".gz",
+    ".xz",
+    ".bz2",
+]
+
+
+def _parse_packaging_version(version: str) -> vn.Version | None:
+    """Parse packaging version."""
+    result = None
+    try:
+        v = vn.parse(version)
+        # do not support post releases of prereleases etc.
+        if not (v.pre and (v.post or v.dev or v.local)):
+            result = v
+    except vn.InvalidVersion:
+        pass
+
+    return result
+
+
+class PyProjectProviderQueryError(Exception):
+    """Error during querying of the PyProjectProvider."""
+
+
+class PyProjectProvider(Protocol):
+    """General provider interface for Python distribution packages."""
+
+    @abc.abstractmethod
+    def get_versions(self, name: str) -> list[vn.Version] | PyProjectProviderQueryError:
+        """Get available package versions.
+
+        Returns an error if no versions are found.
+        """
+
+    @abc.abstractmethod
+    def get_pyproject(
+        self, name: str, version: vn.Version
+    ) -> dict[Any, Any] | PyProjectProviderQueryError:
+        """Get the contents of the pyproject.toml file for the specified version."""
+
+    @abc.abstractmethod
+    def get_hash(
+        self, name: str, version: vn.Version
+    ) -> dict[str, str] | PyProjectProviderQueryError:
+        """Get the sdist hash (sha256 if available) for the specified version."""
+
+
+class PyPIProvider(PyProjectProvider):
+    """Obtains project versions and distribution packages through the PyPI JSON API.
+
+    Various public and private methods of this class cache their return values in order
+    to minimize the number of requests to the provider.
+    """
+
+    def __init__(self, base_url: str = "https://pypi.org/simple/") -> None:
+        """Initialize PyPI Provider with base API url."""
+        if not base_url.endswith("/"):
+            base_url = f"{base_url}/"
+        self.base_url = base_url
+
+    @functools.lru_cache(maxsize=1)  # noqa: B019
```

Comment:
I don't think it is useful to limit the size to just 1 in the methods of this class. Caching some json files shouldn't be a problem resource-wise, and it can help performance anyway, specially when doing transitive/recursive package lookups.
```suggestion
    @functools.lru_cache()  # noqa: B019
```